packages/iov-bns/src/encode.ts

```diff
@@ -102,29 +110,32 @@ export function buildUnsignedTx(tx: UnsignedTransaction): codecImpl.app.ITx {
 }
 
 export function buildMsg(tx: UnsignedTransaction): codecImpl.app.ITx {
-  switch (tx.kind) {
-    case TransactionKind.AddAddressToUsername:
-      return buildAddAddressToUsernameTx(tx);
-    case TransactionKind.Send:
-      return buildSendTx(tx);
-    case TransactionKind.SetName:
-      return buildSetNameTx(tx);
-    case TransactionKind.SwapOffer:
-      return buildSwapOfferTx(tx);
-    case TransactionKind.SwapCounter:
-      return buildSwapCounterTx(tx);
-    case TransactionKind.SwapClaim:
-      return buildSwapClaimTx(tx);
-    case TransactionKind.SwapTimeout:
-      return buildSwapTimeoutTx(tx);
-    case TransactionKind.RegisterBlockchain:
-      return buildRegisterBlockchainTx(tx);
-    case TransactionKind.RegisterUsername:
-      return buildRegisterUsernameTx(tx);
-    case TransactionKind.RemoveAddressFromUsername:
-      return buildRemoveAddressFromUsernameTx(tx);
-    default:
-      throw new Error("Received transacion of unsupported kind.");
+  if (!isBnsTx(tx)) {
+    throw new Error("Transaction is not a BNS transaction");
+  }
+
+  if (isAddAddressToUsernameTx(tx)) {
```

Comment:
Switch does look cleaner, but I see why you use this.
Same type safety in any case.

Reply:
Unfortunatly, TypeScript does not ensure switches are exhaustive on a language level. If it did, the switch gave us more safety. [There are hacks around that](https://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript) but nothing really elegant.

We need the type checkers like `isAddAddressToUsernameTx` anyway because they check `UnsignedTransaction`s not only `BnsTx`s but this is a feature we do not use in this code block.

So this is mostly a readability thing and the switch indeed looked better.